### PR TITLE
[Telemetry] proptest for parsers.rs

### PR DESCRIPTION
**Pull request by ronanM**
I have added `proptest` for parsing `BootMessage` (but I don't understood the `device_id` logic).

If you found that useful, i could continue with other messages.

### CODE REVIEW

**Comment by BlackYoup** [comment]
Hi!

I didn't know proptest, this seems nice. From what I understand, it will test multiple scenarios based on the format you give and will change parts of it each time?

> but I don't understood the device_id logic

I may be mistaken but each makair device will have its own ID which is given in the device_id field. You can find more information in this issue: https://github.com/makers-for-life/makair/issues/120

I'll let @dsferruzza rule on this one but it seems good for me (maybe just remove the commented `             // println!("{:?}",&msg);`)

File: src/software/telemetry/src/parsers.rs [hunk 1]
```diff
@@ -267,17 +267,68 @@ named!(pub parse_telemetry_message<TelemetryMessage>, alt!(
...
+                version: version.to_string(),
+                device_id: format!("{}-{}-{}", device_id1, device_id2, device_id3),
+                systick: systick,
+                mode: mode,
+                value128: value128,
+            };
+
+            // println!("{:?}",&msg);
+
+            // Need to be consistent with sendBootMessage() defined in src/software/firmware/srcs/telemetry.cpp .
```
Comment (dsferruzza) [inline src/software/telemetry/src/parsers.rs:311]:
```suggestion
            // This needs to be consistent with sendBootMessage() defined in src/software/firmware/srcs/telemetry.cpp
```

File: src/software/telemetry/src/parsers.rs [hunk 2]
```diff
@@ -267,17 +267,68 @@ named!(pub parse_telemetry_message<TelemetryMessage>, alt!(
...
+            device_id2 in (0u32..),
+            device_id3 in (0u32..),
+            systick in (0u64..),
+            mode in mode_strategy(),
+            value128 in (0u8..)
+            ) {
+            let msg = BootMessage {
+                version: version.to_string(),
+                device_id: format!("{}-{}-{}", device_id1, device_id2, device_id3),
+                systick: systick,
```
Comment (dsferruzza) [inline src/software/telemetry/src/parsers.rs:306]:
```suggestion
                systick,
```

File: src/software/telemetry/src/parsers.rs [hunk 3]
```diff
@@ -267,17 +267,68 @@ named!(pub parse_telemetry_message<TelemetryMessage>, alt!(
...
+                mode: mode,
```
Comment (dsferruzza) [inline src/software/telemetry/src/parsers.rs:307]:
```suggestion
                mode,
```

File: src/software/telemetry/src/parsers.rs [hunk 4]
```diff
@@ -267,17 +267,68 @@ named!(pub parse_telemetry_message<TelemetryMessage>, alt!(
...
+                value128: value128,
```
Comment (dsferruzza) [inline src/software/telemetry/src/parsers.rs:308]:
```suggestion
                value128,
```

File: src/software/telemetry/src/parsers.rs [hunk 5]
```diff
@@ -267,17 +267,68 @@ named!(pub parse_telemetry_message<TelemetryMessage>, alt!(
...
-        assert_eq!(nom::dbg_dmp(boot, "boot")(input), Ok((&[][..], expected)));
+    fn flat(v: &[&[u8]]) -> Vec<u8> {
+        v.iter().flat_map(|a| a.iter()).map(|v| *v).collect()
+    }
+
+    fn mode_strategy() -> impl Strategy<Value = Mode> {
+        prop_oneof![
+            Just(Mode::Production),
+            Just(Mode::Qualification),
+            Just(Mode::IntegrationTest)
```
Comment (dsferruzza) [inline src/software/telemetry/src/parsers.rs:280]:
```suggestion
            Just(Mode::IntegrationTest),
```

File: src/software/telemetry/src/parsers.rs [hunk 6]
```diff
@@ -267,17 +267,68 @@ named!(pub parse_telemetry_message<TelemetryMessage>, alt!(
...
+        #[test]
+        fn test_boot_message_parser(
+            version in ".*",
+            device_id1 in (0u32..),
+            device_id2 in (0u32..),
+            device_id3 in (0u32..),
+            systick in (0u64..),
+            mode in mode_strategy(),
+            value128 in (0u8..)
+            ) {
```
Comment (dsferruzza) [inline src/software/telemetry/src/parsers.rs:302]:
```suggestion
        ) {
```

**Comment by dsferruzza** [comment]
@ronanM I just noticed that clippy rises several warnings; could you have a look?

I think you just need to have [clippy](https://github.com/rust-lang/rust-clippy#usage) installed and run `cargo clippy --all-targets --all-features`.

**Comment by ronanM** [comment]
I think that all `TelemetryMessage` messages are tested.

File: src/software/telemetry/src/parsers.rs [hunk 7]
```diff
@@ -267,17 +267,341 @@ named!(pub parse_telemetry_message<TelemetryMessage>, alt!(
...
+    proptest! {
+         #[test]
+         fn test_boot_message_parser(
+             version in ".*",
+             device_id1 in (0u32..),
+             device_id2 in (0u32..),
+             device_id3 in (0u32..),
+             systick in (0u64..),
+             mode in mode_strategy(),
+             value128 in (0u8..)
```
Comment (dsferruzza) [inline src/software/telemetry/src/parsers.rs:336]:
```suggestion
             value128 in (0u8..),
```

File: src/software/telemetry/src/parsers.rs [hunk 8]
```diff
@@ -267,17 +267,341 @@ named!(pub parse_telemetry_message<TelemetryMessage>, alt!(
...
+                 &[msg.patient_valve_position],
+                 b"\t",
+                 &[msg.blower_rpm],
+                 b"\t",
+                 &[msg.battery_level],
+                 b"\n",
+             ]);
+
+             let expected = TelemetryMessage::DataSnapshot(msg);
+             assert_eq!( nom::dbg_dmp(data_snapshot, "data_snapshot")(input), Ok((&[][..], expected)) );
```
Comment (dsferruzza) [inline src/software/telemetry/src/parsers.rs:459]:
```suggestion
             assert_eq!(nom::dbg_dmp(data_snapshot, "data_snapshot")(input), Ok((&[][..], expected)));
```

File: src/software/telemetry/src/parsers.rs [hunk 9]
```diff
@@ -267,17 +267,341 @@ named!(pub parse_telemetry_message<TelemetryMessage>, alt!(
...
+                 &[msg.current_alarm_codes.len() as u8],
+                 &msg.current_alarm_codes,
+                 b"\t",
+                 &[msg.previous_alarm_codes.len() as u8],
+                 &msg.previous_alarm_codes,
+                 b"\n",
+             ]);
+
+             let expected = TelemetryMessage::MachineStateSnapshot(msg);
+             assert_eq!( nom::dbg_dmp(machine_state_snapshot, "machine_state_snapshot")(input), Ok((&[][..], expected)) );
```
Comment (dsferruzza) [inline src/software/telemetry/src/parsers.rs:530]:
```suggestion
             assert_eq!(nom::dbg_dmp(machine_state_snapshot, "machine_state_snapshot")(input), Ok((&[][..], expected)));
```

File: src/software/telemetry/src/parsers.rs [hunk 10]
```diff
@@ -267,17 +267,341 @@ named!(pub parse_telemetry_message<TelemetryMessage>, alt!(
...
+                 &msg.expected.to_be_bytes(),
+                 b"\t",
+                 &msg.measured.to_be_bytes(),
+                 b"\t",
+                 &msg.cycles_since_trigger.to_be_bytes(),
+                 b"\n",
+             ]);
+
+             let expected = TelemetryMessage::AlarmTrap(msg);
+             assert_eq!( nom::dbg_dmp(alarm_trap, "alarm_trap")(input), Ok((&[][..], expected)) );
```
Comment (dsferruzza) [inline src/software/telemetry/src/parsers.rs:604]:
```suggestion
             assert_eq!(nom::dbg_dmp(alarm_trap, "alarm_trap")(input), Ok((&[][..], expected)));
```

**Comment by dsferruzza** [comment]
Thanks @ronanM!

**Pull request merged**